static/js/lib/ckeditor/plugins/ResourceLinkMarkdownSyntax.ts

```diff
@@ -1,64 +1,105 @@
 
-import {
-  RESOURCE_LINK_CKEDITOR_CLASS,
-  RESOURCE_LINK
-} from "@mitodl/ckeditor5-resource-link/src/constants"
 import { Shortcode, escapeShortcodes } from "./util"
 import { turndownService } from "../turndown"
-
-export const encodeShortcodeArgs = (...args: (string | undefined)[]) =>
-  encodeURIComponent(JSON.stringify(args))
-
-const decodeShortcodeArgs = (encoded: string) =>
-  JSON.parse(decodeURIComponent(encoded))
+import { RESOURCE_LINK_CONFIG_KEY } from "./constants"
 
 const RESOURCE_LINK_SHORTCODE_REGEX = Shortcode.regex("resource_link", true)
 
+const queryKeys = {
+  uuid:   "ocw_resource_link_uuid",
+  suffix: "ocw_resource_link_suffix"
+}
+
 /**
- * Class for defining Markdown conversion rules for Resource links
+ * Class for defining Markdown conversion rules for Resource links.
+ *  - Converts from markdown to HTML when populating CKEditor
+ *  - Converts from HTML to markdown when saving
  *
- * These are stored in Markdown like this:
+ * Resource links are stored in markdown as:
  *
  * ```md
- * {{% resource_link AUUIDUNLIKEANYOTHER "Here's a link to my resource" %}}
+ * {{% resource_link "some-uuid" "Title of Link" "optional-link-suffix" %}}
  * ```
  *
- * The first argument is the uuid of the resource to which we're linking, and
- * the second argument is that text that should be rendered inside of the link.
+ * and in HTML as:
+ * ```html
+ * <a href="{previewURL} \
+ *  ?ocw_resource_link_uuid={some-uuid} \
+ *  &ocw_resource_link_suffix={optional-link-suffix}"
+ * > Title of Link </a>
+ * ```
+ * Where:
+ *  - {previewURL} portion is ONLY used for previewing the resource in CKEditor
+ *    and is generated by a template specified in CKEditor's `resource-link`
+ *    configuration item.
  *
- * The ResourceEmbed plugin itself is provided via our fork of CKEditor's
- * 'link' plugin.
  */
 export default class ResourceLinkMarkdownSyntax extends MarkdownSyntaxPlugin {
+  hrefTemplate: string
+
+  static get pluginName(): string {
+    return "ResourceLinkMarkdownSyntax"
+  }
+
   constructor(editor: Editor) {
     super(editor)
+    this.hrefTemplate = editor.config.get(RESOURCE_LINK_CONFIG_KEY).hrefTemplate
+    this.validateConfig()
   }
 
-  static get pluginName(): string {
-    return "ResourceLinkMarkdownSyntax"
+  private validateConfig() {
+    invariant(this.hrefTemplate !== undefined, "hrefTemplate is undefined")
+    try {
+      this.makeResourceLinkHref("fake-uuid", "fake-fragment")
+    } catch (err) {
+      console.error("The hrefTemplate is invalid:")
+      throw err
+    }
+  }
+
+  isResourceLinkHref = (href?: string): boolean => {
+    if (!href) return false
+    const url = new URL(href)
+    return url.searchParams.has(queryKeys.uuid)
+  }
+
+  makeResourceLinkHref = (uuid: string, suffix = "") => {
+    const href = new URL(this.hrefTemplate.replace(/:uuid/g, uuid))
+    href.searchParams.set(queryKeys.uuid, uuid)
+    href.searchParams.set(queryKeys.suffix, suffix)
```

Comment:
I don't love storing these on query parameters, but I do think it's better than forking the Link plugin.

I would prefer to store these on data-attributes, which is what our fork was doing. I investigated whether we could do this without forking as described here: https://stackoverflow.com/a/51365516/2747370. That approach does work to add data attributes to the link. However, it has some issues:
1. deleting the link does not delete the data attributes
2. inserting text at the end of the link can be problematic

(2) might sound especially weird. But CKEditor's Link plugin treats adding text at the end of a link as a special case because there is ambiguous add 
- new link text at the end of the link?
- new non-link text at the end of the link?
In CKEditor-speak, when you insert text at the end of the link, should it have the `linkHref` attribute?

Anyway, CKEditor jumps through some hoops to support both of the cases above, which severely complicates adding "extra" link attributes. *CKEditor does have a mechanism, [decorators](https://ckeditor.com/docs/ckeditor5/latest/features/link.html#custom-link-attributes-decorators), for custom link attributes. But it's not flexible enough for what we want. E.g., the attribute values cannot be specified by the user; since we want the attribute value to be a uuid, that's problematic.)